app/uk/gov/hmrc/helptosavestub/controllers/NSIController.scala

```diff
@@ -107,6 +110,87 @@ object NSIController extends BaseController with Logging {
 
   private val ninoStatusRegex = """AS(\d{3}).*""".r
 
+  def queryAccount(resource: String): Action[AnyContent] = Action { implicit request ⇒
+    val queryString = request.rawQueryString
+    resource match {
+      case "account"      ⇒ handleAccountQuery(queryString)
+      case "transactions" ⇒ handleTxnQuery(queryString)
+      case "messages"     ⇒ handleMessagesQuery(queryString)
+      case _              ⇒ BadRequest
+    }
+  }
+
+  private def handleAccountQuery(rawQueryString: String): Result = {
+
+    extractParams(rawQueryString)
+      .fold(error ⇒ {
+        logger.warn("[Handle Account Query] error occurred when trying to extract params")
+        BadRequest
+      },
+        map ⇒
+          validateParams(map).fold(
+            error ⇒ {
+              logger.warn("[Handle Account Query] invalid params")
+              Ok(Json.toJson(error))
+            },
+            nino ⇒ Ok(Json.toJson(getAccountByNino(nino)))
+          )
+      )
+  }
+
+  private def handleTxnQuery(rawQueryString: String): Result = ???
+
+  private def handleMessagesQuery(rawQueryString: String): Result = ???
+
+  private def validateParams(map: Map[String, String]): Either[NSIErrorResponse, NINO] = {
+
+    val nino = map.get("nino")
+    val version = map.get("version")
+    val systemId = map.get("systemId")
+    val correlationId = map.get("correlationId")
+
+    version match {
+      case Some(v) ⇒
+        if (v =!= "1.0") {
+          Left(NSIErrorResponse.unsupportedVersionResponse)
+        } else {
+          systemId match {
+            case Some(s) ⇒
```

Comment:
using cats validation we could avoid doing these checks one by one and also return all validation errors at once. This is what `ValidatedNel` (Nel = non-empty list) would give us - I can go through it with you if you like.

Reply:
ok kool, Suresh has written something out